### PR TITLE
Parse partially typed status names in song select filter query

### PR DESCRIPTION
**Pull request by hlysine**
Besides `Enum.TryParse`, the parser now also checks if any of the status names starts with the input string. This allows users to type `status=r` instead of `status=ranked`, matching stable behavior.

### CODE REVIEW

File: osu.Game/Screens/Select/FilterQueryParser.cs [hunk 1]
```diff
@@ -120,6 +119,14 @@ private static Operator parseOperator(string value)
         private static bool tryParseInt(string value, out int result) =>
             int.TryParse(value, NumberStyles.None, CultureInfo.InvariantCulture, out result);
 
+        private static bool tryParseEnum<TEnum>(string value, out TEnum result) where TEnum : struct
+        {
+            if (Enum.TryParse(value, true, out result)) return true;
+
+            string status = Enum.GetNames(typeof(TEnum)).FirstOrDefault(name => name.StartsWith(value, true, CultureInfo.InvariantCulture));
+            return Enum.TryParse(status, true, out result);
```
Comment (frenzibyte) [inline osu.Game/Screens/Select/FilterQueryParser.cs:127]:
```suggestion
            value = Enum.GetNames(typeof(TEnum)).FirstOrDefault(name => name.StartsWith(value, true, CultureInfo.InvariantCulture));
            return Enum.TryParse(value, true, out result);
```

**Review by frenzibyte — CHANGES_REQUESTED** [review]

**Comment by bdach** [comment]
This will work for this one status case, as the enum names are unambiguous (no common prefixes). But I'm not sure about applying it generally like this. Especially so that using `Enum.GetNames` like this will likely incur allocs for each filter execution.

**Comment by Henry-YSLin** [comment]
The function is currently only used by status. I just wrote it in a generic way in case there are other applicable cases in the future.

As for the allocs, should I cache `Enum.GetNames` in a private field then? I don't feel like this is a particularly hot path though.

**Comment by smoogipoo** [comment]
Don't do it. As you've said, it's not a hot path.

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**